atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ManyClientPoolingContainer.java

```diff
@@ -117,8 +121,8 @@ public synchronized void shutdownPooling() {
 
     @Override
     protected PoolingContainer<Client> delegate() {
-        List<PoolingContainer<Client>> list = containers;
-        return list.get(random.nextInt(list.size()));
+        ImmutableList<PoolingContainer<Client>> list = containers; // volatile read
```

Comment:
`containers` is a `volatile` reference, and we do not want to incur the cost of locking here as `delegate()` is a frequently accessed method under heavy concurrent load, so we do a volatile read of the immutable list reference to a local var, then get a random element from the immutable list.